File: app/src/main/java/com/example/testhexagongame/tiles/tile/Shape/Shape.java
package com.example.testhexagongame.tiles.tile.Shape;

import com.example.testhexagongame.tiles.tile.Box;

import java.util.ArrayList;
import java.util.HashMap;

public abstract class Shape<T extends Shape, U, Sides> {
    private final HashMap<String, Box<T, U, Sides>> sides = new HashMap<>();
    protected final ArrayList<Sides> sidesRules = new ArrayList<>();
    protected final ArrayList<Integer> rotations = new ArrayList<>();
    private Integer currentRotation = 0;

    public Shape() {
        setAdjacentRules();
        setRotationRules();
    }

    protected abstract void setAdjacentRules();
    protected abstract void setRotationRules();

    public void setRotation(Integer rotation) {
        if (!rotations.contains(rotation)) return;
        currentRotation = rotation;
    }

    public Integer getCurrentRotation() {
        return currentRotation;
    }

    public void setAdjacent(String side, Box<T, U, Sides> node) {
        if (!sidesRules.contains(side)) return;
        sides.put(side, node);
    }

    public Box<T, U, Sides> getSide(String side) {
        return sides.get(side);
    }

    public ArrayList<Box<T, U, Sides>> getAdjacents() {
        return new ArrayList<>(sides.values());
    }
}

File: app/src/main/java/com/example/testhexagongame/tiles/tile/Shape/Shape.kt
package com.example.testhexagongame.tiles.tile.Shape

import com.example.testhexagongame.tiles.tile.Box

abstract class Shape() {
    protected val sides = mutableListOf<String>()
    protected var currentRotation = 0
    protected val rotations = mutableListOf<Int>()
    private val adjacent = HashMap<String, Box<Shape>>()

    init {
        this.setAdjacent()
        this.setRotationRules()
    }

    protected abstract fun setAdjacent()

    protected abstract fun setRotationRules()

    fun addAdjacent(side: String, piece: Box<Shape>) {
        if (!sides.contains(side)) return
        adjacent[side] = piece
    }

    fun getAdjacent(side: String): Box<Shape>? {
        if (!sides.contains(side)) return null
        return adjacent[side]
    }

    fun setRotation(rotation: Int) {
        if (!rotations.contains(rotation)) return
        currentRotation = rotation
    }

    fun getRotation(): Int {
        return currentRotation
    }
}